### DISCUSSION TITLE
floating  edge with a markerEnd not pointing on a custom nodes  shape (e.g circle)

### DISCUSSION THREAD

**Discussion by mahdisoultana** [xyflow/xyflow#2296] [General]
### Describe the Bug

we have a problem when we want to make a floating edge with MarkerEnd (arrow) pointed on a node shape for example a circle , the edge don't keep pointing on the node specially when the edge point on the corners of the node 

### Your Example Website or App

https://codesandbox.io/s/sandpack-project-forked-ocw82p?file=/ButtonEdge.js

### Steps to Reproduce the Bug or Issue

1-create a floating edge 
2-create custom node with shape (circle)
3-add initial data nodes and edges 
4- drag the nodes 
5- you see the bug of the edges arrow markers not pointing correctly to the node

### Expected behavior

as a user i expected to see the edge with the arrow pointing on the node without  weird  spacing but i'm seeing the edge not pointing on the node especially  at the corners 

### Screenshots or Videos

![ice_screenshot_20220712-182633](https://user-images.githubusercontent.com/67097548/178555189-20e51a2a-847a-4b61-980a-0f9dd175a8f1.png)

### Platform

- OS: [e.g. macOS, Windows, Linux]
- Browser: [e.g. Chrome, Safari, Firefox]
- Version: [e.g. 91.1]

**Comment by Ipicky22** [comment]
Hi @Mahdi-Soultana !

I tried to solve your issue and I found this solution.
I based myself on CodeSandBox link.
Documentation Custom Edges : https://reactflow.dev/docs/examples/edges/custom-edge/

First, the custom node : 

```javascript
import { Handle, Position } from "react-flow-renderer";
function Node() {
  return (
    <>
      <Handle type="target" position={Position.Top} id="target-node" />
      <article
        style={{
          width: "100px",
          height: "100px",
          borderRadius: "10em",
          backgroundColor: "#055C9D",
          display: "flex",
          justifyContent: "center",
          alignItems: "center"
        }}
      >
        <p style={{ color: "white" }}>Generator</p>
      </article>
      <Handle type="source" position={Position.Bottom} id="source-node" />
    </>
  );
}

export default Node;
```
I took the handles out of the article, it even works with the handles in it, but compared to what the documentation says, I think it's better to take them out. I framed the whole thing with a fragment.

Secondly, the custom edge : 

```javascript
import React from "react";
import { getEdgeCenter } from "react-flow-renderer";
import { getBezierPath } from "react-flow-renderer";
import { svgTextBg } from "./utils.js";

import "./index.css";

const foreignObjectSize = 40;

const onEdgeClick = (evt, id) => {
  evt.stopPropagation();
  alert(`remove ${id}`);
};

export default function CustomEdge({
  id,
  style = {},
  markerEnd,
  sourceX,
  sourceY,
  targetX,
  targetY,
  sourcePosition,
  targetPosition
}) {
  const edgePath = getBezierPath({
    sourceX,
    sourceY,
    sourcePosition,
    targetX,
    targetY,
    targetPosition
  });

  const [edgeCenterX, edgeCenterY] = getEdgeCenter({
    sourceX: sourceX,
    sourceY: sourceY,
    targetX: targetX,
    targetY: targetY
  });

  svgTextBg();
  return (
    <>
      <path
        id={id}
        style={style}
        className="react-flow__edge-path"
        d={edgePath}
        markerEnd={markerEnd}
      />

      <foreignObject
        width={foreignObjectSize}
        height={foreignObjectSize}
        x={edgeCenterX - foreignObjectSize / 2}
        y={edgeCenterY - foreignObjectSize / 2}
        className="edgebutton-foreignobject"
        requiredExtensions="http://www.w3.org/1999/xhtml"
      >
        <body>
          <button
            className="edgebutton"
            onClick={(event) => onEdgeClick(event, id)}
          >
            R
          </button>
        </body>
      </foreignObject>
    </>
  );
}
```

I removed all ```useStore```, I didn't understand what you wanted to do with it.
Based on the documentation, I modified the customEdge props, and I calculated the edgePath with ```getBezierPath``` and the new props. And it works !

I deleted ```sx, sy, tx, ty, path``` variables, with the new props, they become useless.

More detail, it works without the following : 
And finally, in your App file, you don't use the onConnect function provided by the documentation.

Yours : 
```javascript
const onConnect = (params) => setEdges((eds) => addEdge({ ...params, type: "buttonedge" }, eds));
```

Documentation :
```javascript
const onConnect = useCallback(
    (params) => setEdges((eds) => addEdge(params, eds)),
    []
  );
```

Hope to have helped you :)

**Comment by mahdisoultana** [comment]
Hello @Ipicky22 thank you very much for your response i appreciate it ,
this result is acceptable but not fulfilled the use case , the edge is not floating around the circle , what i want is floating edge like you see in the doc but keep pointing where ever we drag the node the edge position should dynamically change the position where it pointed around the node the this why we use floating edge the issue that that i'm facing here with floating edge and has MarkerEnd is when we drag the node at corners we have weird spacing like you see in the image above ?

you can see the floating edge here and why i used ``` useStore,and sx , sy , tx ,ty ```
as show in the documentation floating edge and it's sandbox
https://reactflow.dev/docs/examples/edges/floating-edges/
https://codesandbox.io/s/ynwnhb?file=/FloatingEdge.js:477-492&from-sandpack=true 

i hope you understand better my case , and thank you again for your time 

 try moving the nodes and see the spaces in the corners of the node and this is what the problem with the floating edge and custom node shape circle .
 https://codesandbox.io/s/sandpack-project-forked-ocw82p?file=/ButtonEdge.js

**Comment by Ipicky22** [comment]
Hi @Mahdi-Soultana, sorry, I didn't understand your problem.
I tried to solve it, but without success. But I nevertheless found an approach that could work, with knowledge of math better than mine.

In your ButtonEdge, you can modifie the value of sourceX, sourceY, targetX, targetY.

```javascript
 const { sx, sy, tx, ty, sourcePos, targetPos } = getEdgeParams(
    sourceNode,
    targetNode
  );


  const d = getBezierPath({
    sourceX: sx,
    sourceY: sy,
    sourcePosition: sourcePos,
    targetPosition: targetPos,
    targetX: tx,
    targetY: ty
  });
```

It would be necessary to find a calculation which makes it possible to stick to the rounding of the circle. I got some results by changing the value of targetX and targetY. But nothing that allows you to make a whole quarter or half circle.

If you manage to find the calculation, I want it :)

**Comment by mahdisoultana** [comment]
@Ipicky22  yes sure you got it , we need a math equation for solving this problem
